nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+====invoke====
+Execute a contract invocation.
+
+
+'''Method Interface'''
+
+<pre>
+function invoke({
+  scriptHash: string,
+  operation: string,
+  args: Argument[],
+  attachedAssets?: AttachedAssets,
+  fee?: string,
+  assetIntentOverrides?: AssetIntentOverrides,
+  triggerContractVerification?: boolean,
+  network?: string,
+  broadcastOverride?: boolean,
+  txHashAttributes?: TxHashAttribute[],
+}): Promise<InvokeOutput>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface InvokeArgs {
+  scriptHash: string; // script hash of the smart contract to invoke
+  operation: string; // operation on the smart contract to call
+  args: Argument[]; // any input arguments for the operation
+  fee?: string; // (Optional) The parsed amount of network fee (in GAS) to include with transaction
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+  attachedAssets?: AttachedAssets;
+  broadcastOverride?: boolean; // In the case that the dApp would like to be responsible for broadcasting the signed transaction rather than the wallet provider
+
+  assetIntentOverrides?: AssetIntentOverrides;
+  // A hard override of all transaction utxo inputs and outputs.
+  // IMPORTANT: If provided, fee and attachedAssets will be ignored.
+
+  triggerContractVerification?: boolean; // Adds the instruction to invoke the contract verification trigger
+
+  txHashAttributes?: TxHashAttribute[]; // Adds transaction attributes for the "Hash<x>" usage block
+}
+
+interface Argument {
+  type: ArgumentDataType;
+  value: any;
+}
+
+interface TxHashAttribute extends Argument {
+  txAttrUsage: 'Hash1'|'Hash2'|'Hash3'|'Hash4'|'Hash5'|'Hash6'|'Hash7'|'Hash8'|'Hash9'|'Hash10'|'Hash11'|'Hash12'|'Hash13'|'Hash14'|'Hash15';
+}
+
+type ArgumentDataType = 'String'|'Boolean'|'Hash160'|'Hash256'|'Integer'|'ByteArray'|'Array'|'Address';
+
+interface AttachedAssets {
+  [asset: 'NEO'|'GAS']: string;
+}
+// KEY: Asset symbol (only NEO or GAS)
+// VALUE: Parsed amount to attach
+
+interface AssetIntentOverrides {
+  inputs: AssetInput[];
+  outputs: AssetOutput[];
+}
+
+interface AssetInput {
+  txid: string;
+  index: number;
+}
+
+interface AssetOutput {
+  asset: string;
+  address: number;
+  value: string;
+}
+
+</pre>
+
+
+'''Success return value'''
+
+interface InvokeOutput {
+  txid: string; // The transaction ID of the invocation
+  nodeUrl?: string; // The node which the transaction was broadcast to. Returned if transaction is broadcast by wallet provider
+  signedTx?: string; // The serialized signed transaction. Only returned if the broadcastOverride input argument was set to True
+}
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`SEND_ERROR`|`MALFORMED_INPUT`|`CANCELED`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+invoke({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  operation: 'storeData',
```

Comment:
Shouldn't we just receive a vector of `operation` here, and `operation` has: `name`, `parameters`, ... so that a single `invoke` specification will automatically handle multiple invocations too.

Reply:
it seems that the common pattern suggested for all contracts has two arguments, `operation` and `args`. This "best practice" provides a consistent invocation pattern for new developers to internalize. Are there cases of production contracts which do not follow this pattern?

Reply:
I mean, string op + args is the standard indeed. But you can do two invocations simultaneously, one after another, in a single transaction... I think it ahould be definitely supported,although people would usually just do one single. It can be optional too, perhaps.